// @opentelemetry/plugin-pg/build/src/utils.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.patchCallback = exports.handleInvalidQuery = exports.handleTextQuery = exports.handleParameterizedQuery = exports.handleConfigQuery = void 0;
const api_1 = require("@opentelemetry/api");
const enums_1 = require("./enums");
const pg_1 = require("./pg");
function arrayStringifyHelper(arr) {
    return '[' + arr.toString() + ']';
}
// Helper function to get a low cardinality command name from the full text query
function getCommandFromText(text) {
    if (!text)
        return 'unknown';
    const words = text.split(' ');
    return words[0].length > 0 ? words[0] : 'unknown';
}
function getJDBCString(params) {
    const host = params.host || 'localhost'; // postgres defaults to localhost
    const port = params.port || 5432; // postgres defaults to port 5432
    const database = params.database || '';
    return `jdbc:postgresql://${host}:${port}/${database}`;
}
// Private helper function to start a span
function pgStartSpan(tracer, client, name) {
    const jdbcString = getJDBCString(client.connectionParameters);
    return tracer.startSpan(name, {
        kind: api_1.SpanKind.CLIENT,
        attributes: {
            [enums_1.AttributeNames.COMPONENT]: pg_1.PostgresPlugin.COMPONENT,
            [enums_1.AttributeNames.DB_INSTANCE]: client.connectionParameters.database,
            [enums_1.AttributeNames.DB_TYPE]: pg_1.PostgresPlugin.DB_TYPE,
            [enums_1.AttributeNames.PEER_ADDRESS]: jdbcString,
            [enums_1.AttributeNames.PEER_HOSTNAME]: client.connectionParameters.host,
            [enums_1.AttributeNames.PEER_PORT]: client.connectionParameters.port,
            [enums_1.AttributeNames.DB_USER]: client.connectionParameters.user,
        },
    });
}
// Queries where args[0] is a QueryConfig
function handleConfigQuery(tracer, ...args) {
    const argsConfig = args[0];
    // Set child span name
    const queryCommand = getCommandFromText(argsConfig.name || argsConfig.text);
    const name = pg_1.PostgresPlugin.BASE_SPAN_NAME + ':' + queryCommand;
    const span = pgStartSpan(tracer, this, name);
    // Set attributes
    if (argsConfig.text) {
        span.setAttribute(enums_1.AttributeNames.DB_STATEMENT, argsConfig.text);
    }
    if (argsConfig.values instanceof Array) {
        span.setAttribute(enums_1.AttributeNames.PG_VALUES, arrayStringifyHelper(argsConfig.values));
    }
    // Set plan name attribute, if present
    if (argsConfig.name) {
        span.setAttribute(enums_1.AttributeNames.PG_PLAN, argsConfig.name);
    }
    return span;
}
exports.handleConfigQuery = handleConfigQuery;
// Queries where args[1] is a 'values' array
function handleParameterizedQuery(tracer, ...args) {
    // Set child span name
    const queryCommand = getCommandFromText(args[0]);
    const name = pg_1.PostgresPlugin.BASE_SPAN_NAME + ':' + queryCommand;
    const span = pgStartSpan(tracer, this, name);
    // Set attributes
    span.setAttribute(enums_1.AttributeNames.DB_STATEMENT, args[0]);
    if (args[1] instanceof Array) {
        span.setAttribute(enums_1.AttributeNames.PG_VALUES, arrayStringifyHelper(args[1]));
    }
    return span;
}
exports.handleParameterizedQuery = handleParameterizedQuery;
// Queries where args[0] is a text query and 'values' was not specified
function handleTextQuery(tracer, ...args) {
    // Set child span name
    const queryCommand = getCommandFromText(args[0]);
    const name = pg_1.PostgresPlugin.BASE_SPAN_NAME + ':' + queryCommand;
    const span = pgStartSpan(tracer, this, name);
    // Set attributes
    span.setAttribute(enums_1.AttributeNames.DB_STATEMENT, args[0]);
    return span;
}
exports.handleTextQuery = handleTextQuery;
/**
 * Invalid query handler. We should never enter this function unless invalid args were passed to the driver.
 * Create and immediately end a new span
 */
function handleInvalidQuery(tracer, originalQuery, ...args) {
    let result;
    const span = pgStartSpan(tracer, this, pg_1.PostgresPlugin.BASE_SPAN_NAME);
    try {
        result = originalQuery.apply(this, args);
        span.setStatus({ code: api_1.CanonicalCode.OK }); // this will never happen, but set a status anyways
    }
    catch (e) {
        span.setStatus({ code: api_1.CanonicalCode.UNKNOWN, message: e.message });
        throw e;
    }
    finally {
        span.end();
    }
    return result;
}
exports.handleInvalidQuery = handleInvalidQuery;
function patchCallback(span, cb) {
    return function patchedCallback(err, res) {
        if (err) {
            span.setStatus({
                code: api_1.CanonicalCode.UNKNOWN,
                message: err.message,
            });
        }
        else if (res) {
            span.setStatus({ code: api_1.CanonicalCode.OK });
        }
        span.end();
        cb.call(this, err, res);
    };
}
exports.patchCallback = patchCallback;
